### PR TITLE
migrations guidelines

### PR DESCRIPTION
**Pull request by tommasini**

### CODE REVIEW

File: docs/migrations-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Migration and Testing Guidelines
+
+## Overview
+
+This document outlines best practices and guidelines for writing migration scripts and their corresponding test cases. The focus is on ensuring data integrity, handling errors gracefully, and maintaining consistency across mobile app versions.
```
Comment (NicolasMassart) [inline docs/migrations-guidelines.md:5]:
```suggestion
This document outlines best practices and guidelines for writing migration scripts and their corresponding test cases.
The focus is on ensuring data integrity, handling errors gracefully, and maintaining consistency across mobile app versions.
```

File: docs/migrations-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+## Migration Guidelines
+
+1. **Pre-Validation Checks**: Before proceeding with the migration logic, ensure the state to be migrated is valid. Use utility functions like `ensureValidState` to verify the version and integrity of the state.
```
Comment (NicolasMassart) [inline docs/migrations-guidelines.md:9]:
```suggestion
1. **Pre-Validation Checks**: 
    - Ensure the state to be migrated is valid before running the migration logic, 
    - Verify the version and integrity of the state with utility functions like `ensureValidState`.
```

File: docs/migrations-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+2. **Error Handling**: Use `captureException` from `@sentry/react-native` to log errors, which is crucial for diagnosing issues post-migration. Ensure that error messages are descriptive, including the migration number and a clear description of the issue. If an exception is detected, indicating potential data corruption, halt the migration process and return the state as is. This approach prevents further manipulation of potentially corrupted data, emphasizing the need for corrective action before proceeding.
```
Comment (NicolasMassart) [inline docs/migrations-guidelines.md:11]:
```suggestion
2. **Error Handling**: 

The following prevents manipulation of potentially corrupted data, enforcing corrective action before proceeding.

    - Log erros with `captureException` from `@sentry/react-native`, which is crucial for diagnosing issues post-migration,
    - Ensure that error messages are descriptive: include the migration number and a clear description of the issue,
    - If an exception is detected, indicating potential data corruption, halt the migration process and return the intial state, 
```

File: docs/migrations-guidelines.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+3. **State Integrity Checks**: Perform thorough checks on the state's structure and types. Use functions like `isObject` and `hasProperty` to validate the state and its nested properties. Log errors and halt the migration for any inconsistencies to prevent data corruption.
```
Comment (NicolasMassart) [inline docs/migrations-guidelines.md:13]:
```suggestion
3. **State Integrity Checks**: 
    - Checks the state's structure and types carefully,
    - Validate the state and its nested properties using functions like `isObject` and `hasProperty`,
    - Prevent data corruption by halt the migration for any inconsistencies and logging errors.
```

File: docs/migrations-guidelines.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+4. **Avoid Type Casting**: Refrain from using type casting (e.g., `as` keyword in TypeScript) to manipulate the state or its properties. Type casting can mask underlying data structure issues, making the code less resilient to changes in dependencies or data structures. Instead, use type guards and runtime checks to ensure data integrity and compatibility.
```
Comment (NicolasMassart) [inline docs/migrations-guidelines.md:15]:
```suggestion
4. **Avoid Type Casting**: 
    - Do not use type casting (for example, `as` keyword in TypeScript) to manipulate the state or its properties,
    - Type casting can mask underlying data structure issues and make the code less resilient in case of change in dependencies or data structures,
    - Instead, use type guards and runtime checks to ensure data integrity and compatibility.
```

File: docs/migrations-guidelines.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+5. **Return State**: Always return the state at the end of the migration function, whether it was modified or not. This ensures that the migration process completes and the state continues through any subsequent migrations.
```
Comment (NicolasMassart) [inline docs/migrations-guidelines.md:17]:
```suggestion
5. **Return State**: 
    - Always return the state at the end of the migration function, whether it was modified or not,
    - Returning the state ensures that the migration process completes and the state is passed to the next migrations.
```

File: docs/migrations-guidelines.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+## Testing Guidelines
+
+1. **Mocking**: Use `jest.mock` to mock external dependencies, such as `@sentry/react-native`. This isolates the migration test and prevents external side effects.
```
Comment (NicolasMassart) [inline docs/migrations-guidelines.md:21]:
```suggestion
1. **Mocking**: 
    - Mock dependencies with `jest.mock` (for example `@sentry/react-native`),
    - Mocking isolates the migration under test and prevents side effects.
```

File: docs/migrations-guidelines.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+2. **Initial State Setup**: Create an initial state that reflects possible real-world scenarios, including edge cases. Use utilities like `merge` from `lodash` to combine base states with specific test case modifications.
```
Comment (NicolasMassart) [inline docs/migrations-guidelines.md:23]:
```suggestion
2. **Initial State Setup**: 
    - Create an initial state that reflects possible real-world scenarios, including edge cases,
    - if needed, create multiple initial states and use them each in a test for this specific case,
    - Use utilities like `merge` from `lodash` to combine base states with specific test case modifications.
```

File: docs/migrations-guidelines.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+3. **Invalid State Scenarios**: Test how the migration handles invalid states. This includes null values, incorrect types, and missing properties. Ensure that the migration logs the appropriate errors without modifying the state.
```
Comment (NicolasMassart) [inline docs/migrations-guidelines.md:25]:
```suggestion
3. **Invalid State Scenarios**: 
    - Test how the migration handles invalid states, including null values, incorrect types, and missing propertie,
    - Ensure that the migration logs the appropriate errors without modifying and corrupting the state.
```

File: docs/migrations-guidelines.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+4. **Data Synchronization Tests**: For migrations that synchronize data between controllers, write tests that verify the correct synchronization of data. Check for both the presence and accuracy of the synchronized data.
```
Comment (NicolasMassart) [inline docs/migrations-guidelines.md:27]:
```suggestion
4. **Data Synchronization Tests**: 
    - Write tests that verify the correct synchronization of data for migrations that synchronize data between controllers,
    - Check the presence and accuracy of the synchronized data.
```

File: docs/migrations-guidelines.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+5. **Error Assertions**: Verify that errors are logged correctly for invalid states or unexpected conditions. Use `expect` to assert that `captureException` was called with the expected error messages.
```
Comment (NicolasMassart) [inline docs/migrations-guidelines.md:29]:
```suggestion
5. **Error Assertions**: 
    - Verify that errors are logged correctly for invalid states or unexpected conditions,
    - Use `expect` to assert that `captureException` was called with the expected error messages.
```

File: docs/migrations-guidelines.md [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+6. **Ensure State Immutability**: Always use deep cloning (e.g., `cloneDeep` from `lodash`) on the old state before passing it to the migration function in tests. This practice ensures that the original state object is not mutated during the migration process, preserving the integrity of your test data across different test cases. Mutating the state directly can lead to hard-to-track bugs and false positives or negatives in your tests because subsequent tests might not start with the original state as intended. Deep cloning guarantees that each test case operates on an exact, untouched copy of the state, ensuring test reliability and accuracy.
```
Comment (NicolasMassart) [inline docs/migrations-guidelines.md:31]:
```suggestion
6. **Ensure State Immutability**: 
    - Always use deep cloning on the old state before passing it to the migration function in tests. For example, use `cloneDeep` from `lodash`.
      - Deep cloning preserves the integrity of your test data across different test cases,
      - Ensures the original state object is not mutated during the migration process,
      - guarantees that each test case runs on an correct, clean copy of the state.
    - Never mutate the state directly as this can:
      - lead to hard-to-track bugs and false positives or negatives test results,
      - start subsequent tests with the original state as intended.
```

File: docs/migrations-guidelines.md [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+## Conclusion
+
+Following these guidelines will help ensure that migrations are robust, error-resistant, and maintain data integrity. Testing migrations thoroughly is crucial for identifying potential issues before they affect users. Always aim for clarity and thoroughness in both migration scripts and tests.
```
Comment (NicolasMassart) [inline docs/migrations-guidelines.md:35]:
```suggestion
You will improve migrations robustness by following these guidelines and enforce error-resistance, and maintain data integrity.
You will identify more potential issues with extended testing of migrations before they reach our users.
Always aim for clarity and thoroughness in both migration scripts and tests.
```

**Review by NicolasMassart — COMMENTED** [review]
As discussed, here are some individual suggestions (one for each section) to improve the format (not changing the meaning at all):
- formatting (mostly making it bullet point lists)
- reverting the sentences by putting the action the dev has to do first then how and why doing it.

File: docs/migrations-guidelines.md [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    - Ensure the state to be migrated is valid before running the migration logic, 
+    - Verify the version and integrity of the state with utility functions like `ensureValidState`.
+
+2. **Error Handling**: 
+
+The following prevents manipulation of potentially corrupted data, enforcing corrective action before proceeding.
+
+    - Log erros with `captureException` from `@sentry/react-native`, which is crucial for diagnosing issues post-migration,
+    - Ensure that error messages are descriptive: include the migration number and a clear description of the issue,
+    - If an exception is detected, indicating potential data corruption, halt the migration process and return the intial state,
```
Comment (NicolasMassart) [inline docs/migrations-guidelines.md:20]:
```suggestion
The following prevents manipulation of potentially corrupted data, enforcing corrective action before proceeding.
    - Log erros with `captureException` from `@sentry/react-native`, which is crucial for diagnosing issues post-migration,
    - Ensure that error messages are descriptive: include the migration number and a clear description of the issue,
    - If an exception is detected, indicating potential data corruption, halt the migration process and return the intial state, 
```

File: docs/migrations-guidelines.md [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+- Mock dependencies with `jest.mock` (for example `@sentry/react-native`),
+- Mocking isolates the migration under test and prevents side effects.
+
+2. **Initial State Setup**:
+
+- Create an initial state that reflects possible real-world scenarios, including edge cases,
+- if needed, create multiple initial states and use them each in a test for this specific case,
+- Use utilities like `merge` from `lodash` to combine base states with specific test case modifications.
+
+4. **Invalid State Scenarios**:
```
Comment (Cal-L) [inline docs/migrations-guidelines.md:54]: I realized that we don't currently handle invalid states gracefully. Let's expand on this and mention to backfill data that might be missing or corrupted. For example, we shouldn't just return original state since it will most likely lead to a crash at some point in the app.

Comment (tommasini) [reply docs/migrations-guidelines.md:54]: Hmm but with that we might be not able to fix the original issue, and to do that I'm imagining that we need to return the corrupted data to some default state, that might delete data for the fix we need

Food for thoughts

Comment (Cal-L) [reply docs/migrations-guidelines.md:54]: It won't be perfect right now since we don't have a standardized reconciliation mechanism but to your point, we should be able to fallback to a default state or at least derive it from the original state. Worst case scenario would be to return original state but as we've seen in the past, that almost always leads to unexpected conditions and/or crash.

Comment (tommasini) [reply docs/migrations-guidelines.md:54]: I'm not sure if this would work with any controllers data! Will reach out for we to schedule 15 minutes to work on an example

Comment (Gudahtt) [reply docs/migrations-guidelines.md:58]: Falling back to default state sounds worse than letting the app crash, because we could erase user data that we can't get back. That's why we started this pattern in the extension migrations

Comment (Gudahtt) [reply docs/migrations-guidelines.md:58]: We could consider using a default state fallback and stashing the invalid state somewhere for recovery later. Some risk of bloating state size there, but as long as our default state is small and we don't keep backups too long, that should be OK.

File: README.md [hunk 16]
```diff
@@ -9,6 +9,7 @@ This is a living repository — nothing is set in stone! If you're member of Met
 - [Engineering Principles](./docs/engineering-principles.md)
 - [Guide to Pull Requests](./docs/pull-requests.md)
 - [JavaScript Guidelines](./docs/javascript.md)
+- [Migrations Best Practices](./docs/migrations.md)
```
Comment (Gudahtt) [inline README.md:12]: We should align on the title and file name. Both are different between here and the file at the moment

File: docs/migrations-guidelines.md
Comment (Gudahtt) [inline docs/migrations-guidelines.md:1]: Is this intended to be applicable to both platforms? Many of the steps are mobile-specific at the moment

File: docs/migrations-guidelines.md [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+1. **Pre-Validation Checks**:
+
+- Ensure the state to be migrated is valid before running the migration logic,
+- Verify the version and integrity of the state with utility functions like `ensureValidState`.
+
+2. **Error Handling**:
+
+The following prevents manipulation of potentially corrupted data, enforcing corrective action before proceeding.
+
+- Log erros with `captureException` from `@sentry/react-native`, which is crucial for diagnosing issues post-migration,
```
Comment (Gudahtt) [inline docs/migrations-guidelines.md:24]: Typo:
```suggestion
- Log errors with `captureException` from `@sentry/react-native`, which is crucial for diagnosing issues post-migration,
```

File: docs/migrations-guidelines.md [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+## Migration Guidelines
+
+1. **Pre-Validation Checks**:
+
+- Ensure the state to be migrated is valid before running the migration logic,
+- Verify the version and integrity of the state with utility functions like `ensureValidState`.
+
+2. **Error Handling**:
+
+The following prevents manipulation of potentially corrupted data, enforcing corrective action before proceeding.
```
Comment (Gudahtt) [inline docs/migrations-guidelines.md:22]: This sentence seems to be discussing the "state integrity checks". Maybe we can swap the places of these two guidelines? So that we start by explaining the integrity checks, then follow with explaining how to handle errors resulting from them.

File: docs/migrations-guidelines.md [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+- Ensure that error messages are descriptive: include the migration number and a clear description of the issue,
+- If an exception is detected, indicating potential data corruption, halt the migration process and return the intial state,
+
+3. **State Integrity Checks**:
+
+- Checks the state's structure and types carefully,
+- Validate the state and its nested properties using functions like `isObject` and `hasProperty`,
+- Prevent data corruption by halting the migration on any inconsistencies and logging errors.
+
+5. **Avoid Type Casting**:
```
Comment (Gudahtt) [inline docs/migrations-guidelines.md:34]: Looks like we skipped number 4

Comment (Gudahtt) [inline docs/migrations-guidelines.md:34]: Avoiding type casting is a general rule we have for all TypeScript. Maybe we can omit this because it's a rule for all code.

The migrations do involve more type checking than usual, but perhaps we can explain that in the "State integrity checks" section. I would explain that the input state has type `unknown` because that's how we use TypeScript to validate state integrity before proceeding with a migration.

File: docs/migrations-guidelines.md [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+6. **Return State**:
+
+- Always return the state at the end of the migration function, whether it was modified or not,
+- Returning the state ensures that the migration process completes and the state is passed to the next migrations.
+
+## Testing Guidelines
+
+1. **Mocking**:
+
+- Mock dependencies with `jest.mock` (for example `@sentry/react-native`),
```
Comment (Gudahtt) [inline docs/migrations-guidelines.md:49]: Perhaps we could focus these guidelines to be more specific for migrations. Much of these steps seem applicable to all unit tests

File: docs/migrations-guidelines.md (hunk 15 above)
Comment (Cal-L) [reply docs/migrations-guidelines.md:58]: Makes sense, we'll have to think about this some more before we provide a good solution for gracefully handling. We can leave this out for now and continue brainstorming. @tommasini

File: docs/migrations-guidelines.md
Comment (tommasini) [reply docs/migrations-guidelines.md:1]: I believe this could be used for extension, will write it on a more general way! Thanks for flagging it!

**Comment by tommasini** [comment]
Hey @Gudahtt Thanks for the review!!
I resolved all your comments because I re-wrote these guidelines in a more general and less verbose way, I think it went in the direction of your review comments! Let me know your thoughts

File: docs/migrations-guidelines.md [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+- Ensure that the migration logs the appropriate errors without modifying and corrupting the state.
+
+3. **Error Assertions**:
+
+- Verify that errors are logged correctly for invalid states or unexpected conditions,
+
+4. **Ensure State Immutability**:
+
+- Always use deep cloning on the old state before passing it to the migration function in tests. For example, use `cloneDeep` from `lodash`.
+  - Deep cloning preserves the integrity of your test data across different test cases,
```
Comment (Cal-L) [inline docs/migrations-guidelines.md:57]:
```suggestion
  - Deep cloning preserves the integrity of your test data across different test cases.
```

File: docs/migrations-guidelines.md [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+  - Ensures the original state object is not mutated during the migration process,
```
Comment (Cal-L) [inline docs/migrations-guidelines.md:58]:
```suggestion
  - Ensures the original state object is not mutated during the migration process.
```

File: docs/migrations-guidelines.md [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+  - guarantees that each test case runs on an correct, clean copy of the state.
+- Never mutate the state directly as this can:
+  - lead to hard-to-track bugs and false positives or negatives test results,
```
Comment (Cal-L) [inline docs/migrations-guidelines.md:61]:
```suggestion
  - lead to hard-to-track bugs and false positives or negatives test results.
```

**Review by Cal-L — COMMENTED** [review]
Overall looks good, just a few syntax issues

**Review by Cal-L — APPROVED** [review]
LGTM

**Review by NicolasMassart — DISMISSED** [review]
Looks good to me.

**Comment by gambinish** [comment]
Apologies if this is already a thing, but is there a way to export state logs in the mobile app?

In Extension, you can do this in `Settings > Advanced`

File: docs/migrations-guidelines.md [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+- If an exception is detected, indicating potential data corruption, halt the migration process and return the intial state,
+
+3. **Return State**:
+
+Completes the migration by returning the state, modified or not, ensuring a seamless transition to subsequent migrations.
+
+- Always return the state at the end of the migration function, whether it was modified or not,
+- Returning the state ensures that the migration process completes and the state is passed to the next migrations.
+
+## Testing Guidelines
```
Comment (sethkfman) [inline docs/migrations-guidelines.md:38]: Should we add something to look at the oldest widely adopted version to generate the first set of state for testing?

**Comment by sethkfman** [comment]
> Apologies if this is already a thing, but is there a way to export state logs in the mobile app?
> 
> In Extension, you can do this in `Settings > Advanced`

@gambinish 
Screenshot 2024-11-15 at 10 50 39

**Review by Cal-L — APPROVED** [review]
LGTM

File: docs/migrations-guidelines.md (hunk 24 above)
Comment (tommasini) [reply docs/migrations-guidelines.md:38]: Yes! Let me create another PR introducing that!
